crates/tracing-json/src/lib.rs

```diff
@@ -0,0 +1,222 @@
+    }
+}
+
+struct JsonLayer {
+    collector: TestingTraceCollector,
+}
+impl JsonLayer {
+    pub fn new(collector: TestingTraceCollector) -> Self {
+        Self { collector }
+    }
+}
+
+impl<S> tracing_subscriber::Layer<S> for JsonLayer
+where
+    S: tracing::Subscriber + for<'a> tracing_subscriber::registry::LookupSpan<'a>,
+{
+    fn on_new_span(
+        &self,
+        attrs: &tracing::span::Attributes<'_>,
+        id: &tracing::span::Id,
+        ctx: tracing_subscriber::layer::Context<'_, S>,
+    ) {
+        let mut visitor = JsonVisitor::default();
+        attrs.record(&mut visitor);
+
+        if let Some(span) = ctx.span(id) {
+            // Store the fields in the span for later use
+            let mut extensions = span.extensions_mut();
+            extensions.insert(visitor.fields);
+        }
+    }
+
+    fn on_enter(&self, id: &tracing::span::Id, ctx: tracing_subscriber::layer::Context<'_, S>) {
+        if let Some(span) = ctx.span(id) {
+            let mut span_json = json::json!({
+                "name": span.name(),
+            });
+
+            if let Some(fields) = span.extensions().get::<json::Map<String, json::Value>>() {
+                for (key, value) in fields {
+                    span_json[key] = value.clone();
+                }
+            }
+
+            self.collector.insert(span_json);
+        }
+    }
+}
+
+pub fn with_tracing<F, R>(test_fn: F) -> R
+where
+    F: FnOnce(&TestingTraceCollector) -> R,
+{
+    let mut collector = TestingTraceCollector::new();
+    collector.clear();
+
+    let layer = JsonLayer::new(collector.clone());
+    let subscriber = tracing_subscriber::registry().with(layer);
+    let dispatch = Dispatch::new(subscriber);
+    let _guard = tracing::dispatcher::set_default(&dispatch);
+    test_fn(&collector)
+}
+
+#[derive(Debug)]
+// This is just to provide additional context when a test fails due to invalid tracing
+// Not actually used in code (except for Debug), so we have to allow dead_code
+#[allow(dead_code)]
+pub enum InvalidTrace {
+    TraceLengthMismatch {
+        expected: usize,
+        actual: usize,
+    },
+    InvalidTrace {
+        expected: serde_json::Value,
+        actual: serde_json::Value,
+        index: usize,
+    },
+}
+
+pub fn verify_traces(
+    actual_traces: Vec<json::Value>,
+    expected_traces: Vec<json::Value>,
+) -> Result<(), InvalidTrace> {
+    if actual_traces.len() != expected_traces.len() {
+        return Err(InvalidTrace::TraceLengthMismatch {
+            expected: expected_traces.len(),
+            actual: actual_traces.len(),
+        });
+    }
+
+    for (index, (actual, expected)) in actual_traces
+        .into_iter()
+        .zip(expected_traces.into_iter())
+        .enumerate()
+    {
+        if actual != expected {
+            return Err(InvalidTrace::InvalidTrace {
+                expected,
+                actual,
+                index,
+            });
+        }
```

Comment:
Given that this function is only really used for test code, I suggest relying on `assert_eq!` for _the entire vector_, so that we can get nice diffs. I don't think the assert lib will by default work well with json::value, but we can compare a vector of text-encoded traces and leverage the underlying tooling for diffs.